docs/architecture/adr-016-validator-consensus-key-rotation.md

```diff
@@ -0,0 +1,120 @@
+# ADR 016: Validator Consensus Key Rotation
+
+## Changelog
+
+- 2019 Oct 23: Initial draft
+- 2019 Nov 28: Add key rotation fee
+
+## Context
+
+Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
+key management policy (e.g. https://github.com/tendermint/tendermint/issues/1136). So, we suggest one of the simplest form of
+validator consensus key rotation implementation mostly onto Cosmos-SDK.
+
+## Decision
+
+### Pseudo procedure for consensus key rotation
+
+- create new random consensus key.
+- create and broadcast a transaction with a `MsgRotateConsPubKey` that states the new consensus key is now coupled with the validator operator with signature from the validator's operator key.
+- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
+- start validating with new consensus key.
+- validators using HSM and KMS should update the consensus key in HSM to use the new rotated key after the height `h` when `MsgRotateConsPubKey` committed to the blockchain.
+
+
+### Considerations
+
+- consensus key mapping information management strategy
+    - store history of each key mapping changes in the kvstore.
+    - the state machine can search corresponding consensus key paired with given validator operator for any arbitrary height in a recent unbonding period.
+    - the state machine does not need any historical mapping information which is past more than unbonding period.
+- limits
+    - a validator cannot rotate its consensus key more than `MaxConsPubKeyRotations` time for any unbonding period, to prevent spam.
+    - parameters can be decided by governance and stored in genesis file.
+- key rotation fee
+    - a validator should pay `KeyRotationFee` to rotate the consensus key which is calculated as below
+    - `KeyRotationFee` = `InitialKeyRotationFee` * 2^(number of rotations in `ConsPubKeyRotationHistory` in recent unbonding period)
+- slash module
+    - slash module can search corresponding consensus key for any height so that it can decide which consensus key is supposed to be used for given height.
```

Comment:
the double sign evidence is now handled by the `x/evidence` module, which gets the registered pubkey from the `x/slashing` module. So ultimately the evidence module will have to look up the corresponding consensus address at the given height. 

cc: @alexanderbez

Reply:
thanks for pointing that out! as you said, evidence module gets pubkey from slashing keeper to use for double signing detection. edited as below

_- evidence module
    - evidence module can search corresponding consensus key for any height from slashing keeper so that it can decide which consensus key is supposed to be used for given height._